src/fetching.rs

```diff
@@ -158,6 +163,8 @@ impl<T, C> Fetcher<T, C> {
             // Now we are responsible for fetching the object, reach out to the provider.
             let mut delay = min(MIN_RETRY_DELAY, max_retry_delay);
             let res = loop {
+                // Acquire a permit from the semaphore to rate limit the number of concurrent fetch requests
+                permit.acquire().await;
```

Comment:
If the fetch fails we need to drop the permit explicitly. Should not hold it while sleeping